data/transactions/logic/assembler.go

```diff
@@ -248,57 +249,110 @@ type OpStream struct {
 	HasStatefulOps bool
 }
 
-// createLabel inserts a label reference to point to the next
-// instruction, reporting an error for a duplicate.
-func (ops *OpStream) createLabel(label string) {
-	if ops.labels == nil {
-		ops.labels = make(map[string]int)
+// newOpStream constructs OpStream instances ready to invoke assemble. A new
+// OpStream must be used for each call to assemble().
+func newOpStream(version uint64) OpStream {
+	return OpStream{
+		labels:       make(map[string]int),
+		OffsetToLine: make(map[int]int),
+		typeTracking: true,
+		Version:      version,
+	}
+}
+
+// ProgramKnowledge tracks statically known information as we assemble
+type ProgramKnowledge struct {
+	// list of the types known to be on the value stack, based on specs of
+	// opcodes seen while assembling. In normal code, the tip of the stack must
+	// match the next opcode's Arg.Types, and is then replaced with its
+	// Return.Types. If `deadcode` is true, `stack` should be empty.
+	stack StackTypes
+
+	// bottom is the type given out when known is empty. It is StackNone at
+	// program start, so, for example, a `+` opcode at the start of a program
+	// fails. But when a label or callsub is encountered, `stack` is truncated
+	// and `bottom` becomes StackAny, because we don't track program state
+	// coming in from elsewhere. A `+` after a label succeeds, because the stack
+	// "vitually" contains an infinite list of StackAny.
+	bottom StackType
+
+	// deadcode indicates that the program is in deadcode, so no type checking
+	// errrors should be reported.
```

Comment:
@jannotti Minor typo noticed after approving.  I don't feel strongly, but leave the _suggestion_ in case you want to take it in.

```suggestion
	// errors should be reported.
```